packages/filesystem/src/browser/file-tree/file-tree-widget.ts

```diff
@@ -56,4 +66,140 @@ export class FileTreeWidget extends TreeWidget {
+                    if (target.parentElement) {
+                        target.parentElement.classList.remove(activeDropZone);
+                    }
+                },
+            };
+        }
+        return elementAttrs;
+    }
+
+    protected handleDragEnterEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.add(activeDropZone);
+        }
+    }
+
+    protected handleDragOverEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+    }
+
+    protected handleDropEvent(node: ITreeNode, event: DragEvent): void {
+        event.stopPropagation();
+        event.preventDefault();
+        event.dataTransfer.dropEffect = 'copy'; // Explicitly show this is a copy.
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.remove(activeDropZone);
+        }
+        this.dropHandler(event, node);
+    }
+
+    protected dropHandler(ev: DragEvent, node: ITreeNode) {
+        ev.preventDefault();
+        const data = ev.dataTransfer.items;
+
+        for (let i = 0; i < data.length; i += 1) {
+            if (data[i].kind === 'file') {
+                const entry: WebKitEntry = data[i].webkitGetAsEntry();
+
+                if (entry.isDirectory) {
+                    this.handleDirectory(entry, node.id);
+                } else {
+                    const f: any = data[i].getAsFile();
+                    this.buildAndSendData(node.id, f, f.name);
+                }
+            }
+        }
+    }
+
+    protected handleDirectory(webDirEntry: WebKitEntry, dest: string) {
+
+        const getEntries = (dirEntry: any, destination: string) => {
+            const reader = dirEntry.createReader();
+            reader.readEntries((e: any) => {
+                const results = <WebKitEntry[]>e;
+
+                for (let count = 0; count < results.length; count++) {
+
+                    if (results[count].isDirectory) {
+                        getEntries(results[count], destination);
+                    } else {
+                        const fileEntry = <WebKitFileEntry>results[count];
+                        fileEntry.file(e => {
+                            const f = <File>(e as {});
+                            this.buildAndSendData(destination + dirEntry.fullPath, f, f.name);
+                        });
+                    }
+                }
+            }, (error: any) => {
+                this.logger.error(error.toString());
+            });
+        };
+        getEntries(webDirEntry, dest);
+    }
+
+    protected buildAndSendData(destination: string, file: File, fileName: string) {
+        const fileToCreate = destination + "/" + fileName;
+        this.fileSystem.exists(fileToCreate).then(exist => {
```

Comment:
I don't think that it should be done like that

@svenefftinge maybe we should improve `FileResource` to handle inexisted files on `setContents`, similar to `UserStorageResource`, and call it from here

Reply:
Could you explain what you don't like about using the fs directly?

Reply:
With

```
if (exists()) {
  setContent()
}
```

if the file gets deleted between the exists and the setContent, the setContent will fail.  Allowing setContent to create file can make sure this doesn't happen.

Reply:
>Could you explain what you don't like about using the fs directly?

It is fine to use, my issue is a bit off-topic:  It is strange that one implementation of Resource creates a file on `setContents` and another not.

If it is unifed then here it can be:
```ts
const resource = await this.resourceProvider.get(URI.create(fileToCreate));
resource.setContents(content, { encoding: 'base64' });
```

Reply:
@simark unfortunately, we cannot ensure atomic of such operation even in the backend. I meant about inconsistent semantic of `Resource` implementations and that code can be simplified if they are used here.

Reply:
Opening the file with `fs.open(<path>, "w")` creates the file if it doesn't exist, so it should be safe regardless if the file exists or not, don't you think?